crates/ty_python_semantic/src/types/display.rs

```diff
@@ -15,117 +18,125 @@ use crate::types::function::{FunctionType, OverloadLiteral};
+use crate::types::visitor::TypeVisitor;
 use crate::types::{
     BoundTypeVarInstance, CallableType, IntersectionType, KnownBoundMethodType, KnownClass,
-    MaterializationKind, Protocol, ProtocolInstanceType, StringLiteralType, SubclassOfInner, Type,
-    UnionType, WrapperDescriptorKind,
+    MaterializationKind, Protocol, StringLiteralType, SubclassOfInner, Type, UnionType,
+    WrapperDescriptorKind, visitor,
 };
 use ruff_db::parsed::parsed_module;
 
 /// Settings for displaying types and signatures
-#[derive(Debug, Copy, Clone, Default)]
-pub struct DisplaySettings {
+#[derive(Debug, Clone, Default)]
+pub struct DisplaySettings<'db> {
     /// Whether rendering can be multiline
     pub multiline: bool,
-    /// Whether rendering will show qualified display (e.g., module.class)
-    pub qualified: bool,
+    /// Class names that should be displayed fully qualified
+    /// (e.g., `module.ClassName` instead of just `ClassName`)
+    pub qualified: Rc<FxHashSet<&'db str>>,
 }
 
-impl DisplaySettings {
+impl<'db> DisplaySettings<'db> {
     #[must_use]
-    pub fn multiline(self) -> Self {
+    pub fn multiline(&self) -> Self {
         Self {
             multiline: true,
-            ..self
+            ..self.clone()
         }
     }
 
     #[must_use]
-    pub fn singleline(self) -> Self {
+    pub fn singleline(&self) -> Self {
         Self {
             multiline: false,
-            ..self
+            ..self.clone()
         }
     }
 
     #[must_use]
-    pub fn qualified(self) -> Self {
-        Self {
-            qualified: true,
-            ..self
-        }
-    }
-
-    #[must_use]
-    pub fn from_possibly_ambiguous_type_pair<'db>(
+    pub fn from_possibly_ambiguous_type_pair(
         db: &'db dyn Db,
         type_1: Type<'db>,
         type_2: Type<'db>,
     ) -> Self {
-        let result = Self::default();
-
-        let Some(class_1) = type_to_class_literal(db, type_1) else {
-            return result;
-        };
+        let collector = AmbiguousClassCollector::default();
+        collector.visit_type(db, type_1);
+        collector.visit_type(db, type_2);
 
-        let Some(class_2) = type_to_class_literal(db, type_2) else {
-            return result;
-        };
-
-        if class_1 == class_2 {
-            return result;
+        Self {
+            qualified: Rc::new(
+                collector
+                    .class_names
+                    .borrow()
+                    .iter()
+                    .filter_map(|(name, classes)| (classes.len() > 1).then_some(*name))
+                    .collect(),
+            ),
+            ..Self::default()
         }
+    }
+}
 
-        if class_1.name(db) == class_2.name(db) {
-            result.qualified()
-        } else {
-            result
-        }
+#[derive(Debug, Default)]
+struct AmbiguousClassCollector<'db> {
+    visited_types: RefCell<FxHashSet<Type<'db>>>,
+    class_names: RefCell<FxHashMap<&'db str, FxHashSet<ClassLiteral<'db>>>>,
```

Comment:
Probably not worth the additional work, but instead of allocating a hash set of all discovered `ClassLiteral`s with that name, you could probably introduce an `Option`-like enum
```rs
enum AmbiguityState<'db> {
    Unambiguous(ClassLiteral<'db>),
    Ambiguous,
}
```
When you discover the first class literal, you start with `AmbiguityState::Unambiguous(first_literal)`. And as soon as you discover *another* class literal, you simply flip the state to `Ambiguous`. At that point, we're not interested anymore in which exact class literals contributed to the ambiguity.